doc/PlatformDesignStory.md

```diff
@@ -0,0 +1,572 @@
+a new crate, `libtock_sync`, as not all Tock apps will want it:
+
+```rust
+use libtock_platform::MethodCallback;
+
+pub struct SyncAdapter<AsyncResponse, Syscalls> {
+    response: core::cell::Cell<Option<AsyncResponse>>,
+    syscalls: Syscalls,
+}
+
+impl<AsyncResponse, Syscalls> SyncAdapter<AsyncResponse, Syscalls> {
+    pub const fn new(syscalls: Syscalls) -> SyncAdapter<AsyncResponse, Syscalls> {
+        SyncAdapter { response: core::cell::Cell::new(None), syscalls }
+    }
+}
+
+impl<AsyncResponse, Syscalls: libtock_platform::Syscalls> SyncAdapter<AsyncResponse, Syscalls> {
+    pub fn wait(&self) -> AsyncResponse {
+        loop {
+            match self.response.take() {
+                Some(response) => return response,
+                None => self.syscalls.yieldk(),
+            }
+        }
+    }
+}
+
+impl<AsyncResponse, Syscalls: libtock_platform::Syscalls>
+MethodCallback<AsyncResponse> for SyncAdapter<AsyncResponse, Syscalls> {
+    fn call(&self, response: AsyncResponse) {
+        self.response.set(Some(response));
+    }
+}
+```
+
+## Adding a Unit Test
+
+Before we write the test itself, we should add one more utility to
+`libtock_unittest`. That utility is the `test_component!` macro, which creates a
+thread-local instance of a type and provides `FreeCallback` implementations for
+every `MethodCallback` implementation the type has:
+
+```rust
+#[macro_export]
+macro_rules! test_component {
+    [$link:ident, $name:ident: $comp:ty = $init:expr] => {
+        let $name = std::boxed::Box::leak(std::boxed::Box::new($init)) as &$comp;
+        std::thread_local!(static GLOBAL: std::cell::Cell<Option<&'static $comp>>
+                           = std::cell::Cell::new(None));
+        GLOBAL.with(|g| g.set(Some($name)));
+        struct $link;
+        impl<T> libtock_platform::FreeCallback<T> for $link
+        where $comp: libtock_platform::MethodCallback<T> {
+            fn call(response: T) {
+                GLOBAL.with(|g| g.get().unwrap()).call(response);
+            }
+        }
+    };
+}
+```
+
+We can finally add a unit test to `libtock_console`:
+
+```rust
+#[cfg(test)]
+mod tests {
+    #[test]
+    fn write() {
+        extern crate std;
+
+        use libtock_platform::MethodCallback;
+        use libtock_sync::SyncAdapter;
+        use libtock_unittest::FakeSyscalls;
+        use std::boxed::Box;
+        use std::thread_local;
+        use super::{set_write_buffer, set_write_callback, start_write, WriteCompleted};
+
+        let syscalls: &_ = Box::leak(Box::new(FakeSyscalls::new()));
+        libtock_unittest::test_component![SyncAdapterLink, sync_adapter: SyncAdapter<WriteCompleted, &'static FakeSyscalls>
+                                          = SyncAdapter::new(syscalls)];
+
+        set_write_callback::<_, SyncAdapterLink>(syscalls, 1234);
+        set_write_buffer(syscalls, b"Hello");
+        start_write(syscalls, 5);
+        let response = sync_adapter.wait();
```

Comment:
Four lines, not including setup, seems like a lot for the synchronous printing interface. Do you anticipate some additional library that would wrap this synchronous interface into a single function call? This document seems to suggest that automatically implementing `SyncAdapter` for all asynchronous APIs at once forces all functions for a given driver to only call a single system call, which leads to a not particularly user-friendly API without some additional higher-level library.

For example, it is hard to imagine a console user ever needing `set_write_buffer()` and `start_write()` to be separate functions.

Reply:
I agree that four lines is a lot for a single synchronous print call. There are a lot of details and options here that I didn't mention.

### Why are `set_write_callback` and `set_write_buffer` separate?

This design is intended to support apps that perform all `allow`/`subscribe` calls at initialization time (i.e. before "runtime"), apps that perform all `allow`/`subscribe` at runtime (as this app does), and apps that mix the two approaches. As such, I intentionally kept `allow` and `subscribe` separate, and with `SyncAdapter` they could be called during app initialization rather than at runtime.

That said the console driver currently requires you to re-`allow` the buffer for every write even if the same buffer is re-used. As such, we can safely combine `set_write_buffer` and `start_write` into a single method on console. I'm not sure that this will be the case in Tock 2.0, so I'll probably keep them separate for now. I'll combine them if Tock 2.0's console driver has the same semantics.

### Why are `start_write` and `sync_adapter.wait()` two separate calls?

From reading this document, it would seem that the obvious answer is "`SyncAdapter` doesn't know how to call `start_write`". That answer isn't actually correct: the virtualization adapters (which I did not mention in this document) will need a way to call `start_write` too. In the real implementation, `start_write` will be in a generic trait (e.g. `AsyncFn<Request>`), just as the callbacks are in generic traits in this design. As a result, `SyncAdapter` actually *could* perform the `start_write` call (so instead of `SyncAdapter::wait` we would have `SyncAdapter::call_and_wait`).

The real reason I didn't implement that is that `SyncAdapter` doesn't know that there will be exactly one callback for each request. I omitted error handling in the document, but if `start_write` fails, there may not be a callback. This issue can be mitigated by use of a trait: examine the response of `start_write` to determine whether to wait for a callback. But what about APIs that generate more than one callback for a single start call? An example of this would be a button: you subscribe to receive callbacks for the button then receive one callback each time it is pressed. It is not obvious how `SyncAdapter` can address both cases.

There's probably a reasonable solution here that involves making the traits a bit more complex, but I haven't fully thought it through.

### Okay, what are some of our options here?

For one, we can write a wrapper library that wraps the console driver (and perhaps `SyncAdapter`) in a nice `println!()` function. I think this is your suggestion. This would work, and seems ideal from an API usability perspective, but it requires maintaining the library on a per-driver basis. Regardless, it seems like a good idea to do this for console because the `println!()` macro's formatting capabilities are *really nice*, and we can't create that in a generic way.

We could also expose more methods, some of which combine the basic methods. Console could expose `write_buffer`, a wrapper around `set_write_buffer` and `start_write`. That would remove 1 line from the setup. We could even expose `write_buffer_with_callback`, which would internally call `set_write_callback`, `set_write_buffer`, and `start_write` in sequence. With the `AsyncFn<Request>` trait for virtualization (as mentioned above), we can put these into separate trait implementations -- e.g. `AsyncFn<StartWrite>` and `AsyncFn<WriteBufferWithCallback>`. We can combine this approach with my earlier suggestion of a `SyncAdapter::call_and_wait` method, which would combine all 4 lines into 1. The downside here is the additional interface complexity of having more methods/traits on the console driver, but it would also provide advantages to asynchronous apps.

Reply:
Thanks for explaining all that, that all makes sense. Yes I was suggesting that wrapping the driver into a `println!()` or similar would be nice, and am glad to hear you don't think this design precludes doing so in any way.

My main concern is that the stated justification for `SyncAdapter` is "to avoid re-implementing synchronous APIs for every driver", but structuring the API to make this possible leads to synchronous API's that are not particularly user friendly -- looking at the `libtock-c` synchronous API's for example you see that many of the exposed functions call several system calls underneath. If creating user friendly synchronous API's will usually require an additional driver layer above the `SyncAdapter` layer, I wonder if automatically getting a synchronous API via the `SyncAdapter` method is actually that useful (we could just reimplement the synchronous API for each driver as part of making the user-friendly higher level API, and perhaps do so more efficiently than when we are forced to use the automatically created one).

I suspect there might be additional advantages to the `SyncAdapter` approach that I am missing, or that I am underestimating the additional work necessary to reimplement a synchronous API for every driver (perhaps for many simpler drivers a more user-friendly/higher level API is not necessary, so the automatic implementation is genuinely useful)

Reply:
I've been trying to decide between two approaches to `libtock-rs`' design for a while.

The first approach looks like the Tock kernel's design: each asynchronous components has its own client trait and maybe its own API trait (I don't recall whether an API trait is necessary). In this case, each component would need to implement each of the following independently:

1. Something like my `static_component!` macro to link it with its dependencies.
1. Virtualization
1. A synchronous API
1. A `Future`-based API

The second approach is the one I described here. You hide the component's API behind a generic trait (note: I only described the client side, but there would need to be traits for the component's API too). This makes the component itself more difficult use, as there are more traits involved and the specific method names are generic (e.g. `call` rather than `start_write`). However, it allows you to share the `static_component!`/virtualization/synchronou/`Future` code between `libtock-rs` components.

I think the advantage of sharing the `static_component!` and `Future` API code between drivers is a larger advantage than sharing `SyncAdapter`, as that code is more difficult to maintain.

Reply:
Oh, I forgot the other advantage of the trait approach: you can enforce the "only call callbacks from within a kernel callback" rule. To do so, we add a second argument to `FreeCallback::call`:

```rust
// Only libtock_platform can create a CallbackContext. CallbackContext is created for kernel callbacks.
#[derive(Copy)]
pub struct CallbackContext<'c> { _private: () }

pub trait FreeCallback<AsyncResponse> {
    fn call(callback_context: CallbackContext, response: AsyncResponse);
}

pub trait MethodCallback<AsyncResponse> {
    fn call(&self, callback_context: CallbackContext, response: AsyncResponse);
}
```

This will help avoid the "accidental recursion" issues the kernel has experienced.

Reply:
Those reasons are pretty compelling, sounds good to me.

Re: virtualization -- why is virtualization needed in userspace? `libtock-c` does not virtualize any resources in userspace. Per-app virtualization of pretty much all devices is already implemented in the kernel. Is OpenTitan expecting to use sufficiently complex userspace apps that each app may have multiple independent threads of execution requiring the same resources? Can this be expected to be true for *most* resources? (I could see it being true for, say, `console` and `gpio`, but for most resources it seems unlikely?). 

Tock already has two layers of virtualization:
1. Virtualization of a resource between kernel capsules (e.g. `capsules/src/virtual_alarm.rs`)
2. Virtualization of a resource between applications (e.g. `capsules/src/alarm.rs`)

So it is probably worth considering carefully whether a third layer of virtualization is worth the extra overhead for all resources. If multiple independent uses of a resource by a single application is rare, perhaps it would be preferable to require that such an application be split into two that communicate via IPC. Or perhaps only a small few resources should be virtualized in userspace, rather than using some approach to automatically virtualize all resources.

In general, my experience in the Tock kernel is that it is difficult to write low-cost virtualizers without heap allocation.

Reply:
What are we virtulising in userspace?

I still like Hudson's idea of a synchronous API calling multiple calls and hiding complexity from users. The trade off being that we need to implement it for all components, but that doesn't seem that bad.

Reply:
Hmm, I assumed that virtualization would be necessary for the same reason the kernel needs virtualization between capsules. I did not know that `libtock-c` doesn't have virtualization -- that really highlights my bad assumption!

In my [ZST Pointer Async](https://github.com/tock/design-explorations/tree/master/zst_pointer_async#incremental-migration-experiment) doc, I claimed that virtualization is necessary in order to allow apps to incrementally migrate from the `Future`-based API over to the lightweight API. However, I think my claim was too strong: it is important to be able to support different types of clients as different call sights, but I never needed concurrent outstanding requests to a single API.

If having multiple concurrent requests to the same API is that rare then I think it is fine to make apps do the virtualization on their own.

I should probably build a prototype console driver that *doesn't* use the traits that I described in this design, and see how it turns out. Hopefully it doesn't take as long to prototype as this did...

Reply:
On the bright side, not having to find a way to generically implement virtualization will probably make the traits-approach much easier moving forward!

And I don't think you necessarily need to build a prototype without traits, if after considering this feedback you feel pretty strongly that the trait-based approach is preferable I am happy to trust your instincts on this, as you have thought much more about this than most of us.

Reply:
The "preventing accidental recursion" benefit is pretty compelling on its own, IMO

Reply:
When I think it through further, using the full set of traits needed to generically implement virtualization (i.e. traits to start asynchronous operations) would be confusing and result in hard-to-use drivers. I've been focusing most of my design work on the main asynchronous operations, but until I had this example I didn't think about how `allow`/`subscribe`/`command` interact. Having `command` be hidden behind a trait while `allow` and `subscribe` are in member functions would probably lead to a lot of confusion and bugs.

The callback traits (`FreeCallback` and `MethodCallback`) don't have that issue. I also think there is significant value in using `SyncAdapter` in unit tests, even if we custom-write synchronous adapters around drivers for application code to use.

The other issue I'm trying to resolve in this design is the awkward nature of `test_component!`. I currently use a thread local variable to make lifetimes work out. This is because the real Tock system calls cannot store a pointer to the type implementing the callback, but that causes lifetime issues in the unit tests. I'm trying to see if I can fix that by modifying the Syscalls trait.